agent/engine/docker_container_engine.go

```diff
@@ -712,13 +715,24 @@ func metadataFromContainer(dockerContainer *docker.Container) DockerContainerMet
 	return metadata
 }
 
+// CopyDockerEventsWithoutBlocking copies the event from input to output without blocking the input
+func CopyDockerEventsWithoutBlocking(in chan *docker.APIEvents, out chan *docker.APIEvents) {
```

Comment:
I don't understand how this doesn't block.  This just means that you have an effective buffer of `2*containerEventsBufferSize` and if that's full, you'll still block...

Reply:
Even if the `out` is full, the read of `in` won't be block, as it simply reads and creates a go routine to write. In this case these go routines will be blocked until `out` isn't full, but go docker client can keeps writing to `in` so that agent won't miss any event.

In rare case, if the process to create the go routine slows down, which means producer is much faster than consumer, the `in` could be full, that's why I had a buffer size of 100 for `in`. This is the best way I can think of, please suggest.

Reply:
I missed that you're creating a goroutine for *each* event.  I think this is not a good idea; we'll have goroutine explosion and will lose any ordering that was provided by go-dockerclient.

You can implement a non-blocking, infinite (subject to how much memory you have) buffer through appending to a slice.  However, I think the more important question here is to determine what the correct behavior should be.

* Do we rely on every single event being delivered?  We have logic that tolerates event loss and reconciles with explicit `inspect`s.
* What happens when we lose an event? Our reconciliation runs every 10 minutes, so we could have a 10-minute window of inconsistency.
* Do we need every event?  No, there's only a small subset of events that we pay attention to at all.

Reply:
I agree this is not a good idea, but this is the same as how the agent works today. Previously the go routine to send event was created in go docker client code, but now they are created in the agent code.

Currently, the go docker client doesn't grantee the order of the event sent, so I think agent have the tolerance for the order of the event. 

>Do we rely on every single event being delivered? We have logic that tolerates event loss and reconciles with explicit inspects.
What happens when we lose an event? Our reconciliation runs every 10 minutes, so we could have a 10-minute window of inconsistency.

Ten minutes is really a big gap, we probably don't want the task status change report have a 10 minutes delay.

>Do we need every event? No, there's only a small subset of events that we pay attention to at all.

We don't need all these events, but the agent has to do the analysis, currently go docker client doesn't support the filter.

I will try to implement a non-blocking infinite buffer to solve the goroutine explosion problem while keeping the order of the event. Any concerns or suggestions?

Reply:
It might make more sense to filter here so we can discard early.  It also might make more sense to do analysis of expected event volume under high load to determine if we can figure out an appropriate buffer size.